### PR TITLE
Merge the recent Pipeline-related changes

### PR DESCRIPTION
**Pull request by r-xue**
~~Looks like the only conflict is .gitignore.~~ resolved

Once you verify that all functions are still intact, we could merge this PR.
~~Afterwards, I will start to use branching directly from your forked repo.~~ We can do this every once in a while, and keep them from diverging too much.

There could be issues when you merge your local branches to `main`, but we can figure that out together later.

### CODE REVIEW

**Comment by psheehan** [comment]
Excellent, I will take a look at this and test it out in the next few days, and will merge once I've double checked that all remains working.

**Comment by r-xue** [comment]
Sure. I guess the best way to look through the changes might be to check on individual comments (I have short descriptions in the commit message). Meanwhile, feel free to comment on the diff, and I am happy to provide background for individual changes if they look questionable.

**Comment by psheehan** [comment]
Ok, ran through this the first time and largely all looks good. A few thoughts/issues I noticed to possibly think about:

1. I needed to add `sys.path.append("./")` ahead of `from selfcal_helpers import *` to not get an error. Possibly user error, though?
2. When I ran this on a dataset, now all of the content of the casalogger is duplicated in the terminal - is this intentional/what is supposed to happen?
3. Some of the weblog table boundary lines for cells seem to have disappeared - also possibly intentional?

Otherwise this seems to run well, and doesn't appear to have affected the output, so generally looks good to me!

**Comment by r-xue** [comment]
Hi @psheehan, thanks for taking a look at this.

> I needed to add sys.path.append("./") ahead of from selfcal_helpers import * to not get an error. Possibly user error, though?

Yes. This is by design. I would like to avoid copying "scripts" into the working directory. So inserting the path becomes necessary if we don't "install" it as a module:
I had this piece for running the code (test_auto_selfcal.py):
```python
for path in ['somewhere/auto_selfcal']:
    abs_path = os.path.abspath(os.path.expanduser(path))
    sys.path.insert(0, abs_path)
from auto_selfcal import selfcal_workflow
selfcal_workflow()
```
, essentially the same as you did.
Locally, I already put two *.py under `auto_selfcal/`. Then the file structure becomes:
```console
auto_selfcal/*py
auto_selfcal/__init__.py
setup.py
readme.md
```
Eventually, the repo could be installed as a standard Python package (going to the site-package area), or, we just ask the user to do the path hack (better than copying , imho). This will be in the next PR.

> When I ran this on a dataset, now all of the content of the casalogger is duplicated in the terminal - is this intentional/what is supposed to happen?

Does this situation (duplicated logger entries) happen from the latest commit (292bb78baeda331e306f470b5989626c0d4e6330)? There are several circumstances I have to think about, but it's likely that one coming from CASA logger, the other one from Python/logging module. Were you using a CASA version with pipeline installed (e.g. casa-6.4.1-...-pipeline-...-py3.6.tar.xz)

Also, I think I might have introduced a defect: when one runs the script several times in one session, the logger might be registered more than once (therefore additional entries). I will take a look at this.

> Some of the weblog table boundary lines for cells seem to have disappeared - also possibly intentional?

One line in an HTML-code output function got commented out accidentally (while I was trying to experiment using it for the Pipeline weblog), so the HTML table doesn't look right. I did see it locally in a test but haven't paid much attention to the problem yet since Pipeline uses a different routine. Thanks for pointing it out.

---

In summary, I will fix the potential duplicated logging issue and the missed table formatting line, in the next commit. Then I guess we are good to go.

**Comment by psheehan** [comment]
> Does this situation (duplicated logger entries) happen from the latest commit (https://github.com/psheehan/auto_selfcal/commit/292bb78baeda331e306f470b5989626c0d4e6330)? There are several circumstances I have to think about, but it's likely that one coming from CASA logger, the other one from Python/logging module. Were you using a CASA version with pipeline installed (e.g. casa-6.4.1-...-pipeline-...-py3.6.tar.xz)

Yes, this duplication happened from that commit. I was running using the base `casa` version from the cvpost machines, so I don't believe that had the pipeline installed. ~~I will try with `casa-alma`, which I believe should to check this.~~ Edit: just checked with `casa-alma` and I still see this behavior.

**Comment by r-xue** [comment]
The commented weblog-related line has been reversed in the last commit.

About the duplicate entries: since I still can't reproduce it locally, we might have to go through this together and examine how the code was executed in your setup. 

I ruled out some possible causes, but I see another likely reason which is related to how the loggers grab their names from the module path: if the `helper` uses a logger named `auto_selcal.selfcal_helper` and the workflow function somehow is using a logger called `auto_selfcal`, the logging hierarchy might create the situation you see. But these are just my speculations at this moment.

Suppose that the logging issue is not critical at this moment, maybe we can address it later after this PR (then we can get the bulk of early codestyle-related changes out of the way).

**Comment by psheehan** [comment]
Ok, I can confirm that the weblog is now fixed with the most recent commit.
